opentofu: init at v1.6.0-alpha1

gmemstr opened the pull request:
## Description of changes

OpenTofu is a fork of Terraform pre-BSL, so the package is pretty similar to the Terraform one with names swapped around. Happy to take this further if needed but given the current state of opentofu I felt this was probably? okay. That said, I did remove the existing Terraform maintainers as well, but am happy to re-add them if they're good with maintaining opentofu as well.

~~Opentofu is not yet tagged with a version so I've opted to go with the latest commit at the time of creation.~~

Closes https://github.com/NixOS/nixpkgs/issues/253469

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

gmemstr committed:
opentofu: init at v1.6.0-alpha1

gmemstr commented:
> It seems that opentofu can be packaged with terraform.overrideAttrs ...? That way, maintainers don't have to review hundreds of lines of code.

The expectation I have is that opentofu is going to diverge from Terraform in a way that means the overrides added would end up making more sense as  a new package/splitting off. I'm not against merging the two but I'm a little cautious about this.

> Probably tag some of them in this PR?

I left a comment in the associated issue that should hopefully reach their inboxes, but I may as well pull them in here as well @marsam @zowoq @techknowlogick

> Use version = "unstable-2023-XX-XX", according to [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md?rgh-link-date=2023-09-20T15%3A30%3A56Z#package-naming).

Done! :)

sugar700 reviewed (changes requested):
I'm not a fan of the current derivation name `opentofu.opentofu_unstable`. I don't think it's intended, and the name is this long because `callPackage` accesses an object that contains `opentofu_unstable` field. Could we simplify that to just `opentofu`?

xfix commented:
Terraform package mostly did that because in the past we had multiple versions of Terraform, but now that we only have Terraform 1, this is no longer necessary due to [Terraform v1 compatibility promises](https://developer.hashicorp.com/terraform/language/v1-compatibility-promises).

That said, I have no idea what sort of compatibility promises OpenTofu will have, but I think that major breaking changes in a popular programming language are pretty universally understood as a bad idea after Python 3.

I think `opentofu` as a package name (with no namespacing, not `opentofu.opentofu`) sounds right to me, and if branching will be necessary, we could call packages `opentofu_2` or something like that.

Note that `terraform`/`terraform_1` is defined as `(callPackage ../applications/networking/cluster/terraform { }).terraform_1`.

gmemstr commented:
Took me a little bit but managed to get the package out of the namespace and ready for another review.

cafkafk commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+      description =
+        "Tool for building, changing, and versioning infrastructure";
+      homepage = "https://opentofu.org/";
+      changelog = "https://github.com/opentofu/opentofu/blob/${version}/CHANGELOG.md";
```
Currently, I'm pretty sure this evaluates to https://github.com/opentofu/opentofu/blob/unstable-2023-09-20/CHANGELOG.md

It probably should be the `src.revision` instead https://github.com/opentofu/opentofu/blob/922f61d84742804dc4357a90ae56c20a07a80683/CHANGELOG.md

cafkafk reviewed (changes requested):
Hey, really cool that you're doing this, can't wait to have this available.

gmemstr commented:
Bumped the revision and addressed comments from @cafkafk. Also noticed that the way to build the binary changed slightly and the main binary's name changed so updated that.

adamcstephens reviewed (approved):
Result of `nixpkgs-review pr 256307` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

azahi commented:
Result of `nixpkgs-review pr 256307` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

azahi commented:
@ofbord build opentofu opentofu.passthru.tests

gmemstr commented:
Noted the failures and updated the binary name in the wrapping process.

gmemstr commented:
@ofborg build opentofu opentofu.passthru.tests

jocelynthode commented:
Thanks  for your work @gmemstr! OpenTofu finally released their first version: https://github.com/opentofu/opentofu/releases/tag/v1.6.0-alpha1  might be worth it to switch to tags from now on :)

gmemstr commented:
Dangit, of course they did <:) PR updated

adamcstephens reviewed (approved):
Let's get this one merged so future bumps should be quicker. :)

---

Result of `nixpkgs-review pr 256307` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 256307` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>opentofu</li>
  </ul>
</details>

dylanmtaylor reviewed (approved):
I agree with Adam. We should definitely get an initial version in.

azahi commented:
Looks like the tests are now failing because `registry.opentofu.org` is not accessible.

gmemstr commented:
@azahi yeah not sure what's up with that or how we want to proceed, whether it be removing the tests or ignoring them for the time being.

gmemstr commented:
Unless I messed something up while moving around the tests and their inputs. Could be related to that but not 100% sure where I went wrong.

azahi commented:
> @azahi yeah not sure what's up with that or how we want to proceed, whether it be removing the tests or ignoring them for the time being.

The PR looks good and I also want to see this merged. I guess we can just disable checks for now and leave a comment explaining the reasoning. The most important part is to not to forget about all of that later :)

> Could be related to that

I think they broke something on their end. It just returns "Missing Authentication Token".

SuperSandro2000 commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+let
+  package =  buildGoModule rec {
+    pname = "opentofu";
+    version = "v1.6.0-alpha1";
+
+    src = fetchFromGitHub {
+      owner = "opentofu";
+      repo = "opentofu";
+      rev = "${version}";
```
```suggestion
    version = "1.6.0-alpha1";

    src = fetchFromGitHub {
      owner = "opentofu";
      repo = "opentofu";
      rev = "v${version}";
```

SuperSandro2000 commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+    ldflags = [ "-s" "-w" ];
+
+    postConfigure = ''
+        # speakeasy hardcodes /bin/stty https://github.com/bgentry/speakeasy/issues/22
+        substituteInPlace vendor/github.com/bgentry/speakeasy/speakeasy_unix.go \
+          --replace "/bin/stty" "${coreutils}/bin/stty"
+      '';
```
```suggestion
      # speakeasy hardcodes /bin/stty https://github.com/bgentry/speakeasy/issues/22
      substituteInPlace vendor/github.com/bgentry/speakeasy/speakeasy_unix.go \
        --replace "/bin/stty" "${coreutils}/bin/stty"
    '';
```
I am surprised that go is not complaining about this

SuperSandro2000 commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+
+    # https://github.com/posener/complete/blob/9a4745ac49b29530e07dc2581745a218b646b7a3/cmd/install/bash.go#L8
+    postInstall = ''
+        installShellCompletion --bash --name tofu <(echo complete -C tofu tofu)
+      '';
+
+    preCheck = ''
+        export HOME=$TMPDIR
+        export TF_SKIP_REMOTE_TESTS=1
+      '';
```
```suggestion
      installShellCompletion --bash --name tofu <(echo complete -C tofu tofu)
    '';

    preCheck = ''
      export HOME=$TMPDIR
      export TF_SKIP_REMOTE_TESTS=1
    '';
```

SuperSandro2000 commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+    subPackages = [ "./cmd/..." ];
+
+    meta = with lib; {
+      description =
+        "Tool for building, changing, and versioning infrastructure";
```
```suggestion
      description = "Tool for building, changing, and versioning infrastructure";
```

SuperSandro2000 commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+
+  opentofu_plugins_test = let
+    mainTf = writeText "main.tf" ''
+        resource "random_id" "test" {}
+      '';
```
```suggestion
      resource "random_id" "test" {}
    '';
```

SuperSandro2000 commented on pkgs/applications/networking/cluster/opentofu/default.nix:
```diff
@@ -0,0 +1,173 @@
+        resource "random_id" "test" {}
+      '';
+    opentofu = package.withPlugins (p: [ p.random ]);
+    test =
+      runCommand "opentofu-plugin-test" { buildInputs = [ opentofu ]; } ''
+          set -e
+          # make it fail outside of sandbox
+          export HTTP_PROXY=http://127.0.0.1:0 HTTPS_PROXY=https://127.0.0.1:0
+          cp ${mainTf} main.tf
+          tofu init
+          touch $out
+        '';
```
```suggestion
    test = runCommand "opentofu-plugin-test" {
      buildInputs = [ opentofu ];
    } ''
      # make it fail outside of sandbox
      export HTTP_PROXY=http://127.0.0.1:0 HTTPS_PROXY=https://127.0.0.1:0
      cp ${mainTf} main.tf
      tofu init
      touch $out
    '';
```

set -e should already be set here

SuperSandro2000 reviewed (changes requested):
I only correct the first few indentations which where 4 instead of 2 spaces.
